### ISSUE TITLE
Security vulnerabilities in bundled ffmpeg binaries

### ISSUE THREAD

**Issue by kaixxx** [kaixxx/noScribe#108]
### Discussed in https://github.com/kaixxx/noScribe/discussions/106

<div type='discussions-op-text'>

<sup>Originally posted by **tidely** November 25, 2024</sup>
noScribe comes with 3 bundled ffmpeg binaries, to my understanding this is for the sake of not having any outside dependencies from the actual app bundle itself. However it seems the bundled ffmpeg binaries are not updated regularly and are currently on version 6.0. There have been multiple [security vulnerabilities ](https://www.ffmpeg.org/security.html) since this release. In my opinion the ffmpeg should be provided by the system and not bundled with the app. Since every time ffmpeg gets an update, you would need to repackage the app and update it for all users. ffmpeg is very popular and gets updated automatically on linux and macos (apt and brew). At the very least I'd encourage downloading the newest ffmpeg binaries and updating the app bundle.

Bundled ffmpeg

- No external dependency
- 240mb larger repository size
- Need to rebuild and distribute update every time ffmpeg updates

System ffmpeg

- External dependency
- Takes no space in the repository
- ffmpeg gets updated by the system

TLDR: currently used old ffmpeg has [security vulnerabilities ](https://www.ffmpeg.org/security.html)
</div>

**Comment by kaixxx** [comment]
@tidely: Thanks for bringing this up. I've turned this into an issue as a reminder for myself. I will try to find a solution for the next release. Having no external dependencies is still something I would like to keep. But we should bundle the latest binary of ffmpeg and only one, if possible.

**Comment by gernophil** [comment]
Do we really bundle all binaries for every OS?

**Comment by kaixxx** [comment]
> Do we really bundle all binaries for every OS?

No. But ffmpeg comes pre-bundled with some libraries, and we use it directly in noScribe. This is why the binary is included three times (in different places) during the compilation with pyinstaller. Not very efficient...

**Comment by tidely** [comment]
I don't believe any packages in the requirements.txt bundles ffmpeg. I think the only ffmpeg files included happen when noScribe itself is bundled. Since there are platform-specific `.spec` files, it should only get included once. However those are outdated.

I think the best approach would be to make noScribe.py use the system ffmpeg binary when it is available, using `shutil.which()` for example, and use the bundled binaries as a fallback. Additionally we could check the ffmpeg website for updates during compile time, this would mean that ffmpeg gets updated everytime a new noScribe version is released, but the user wouldn't need to download anything extra.

**Comment by kaixxx** [comment]
I took a quick look, and you are right, it's not different packages. One version of ffmpeg is included with the main program, and two with noScribe editor (in different places). I will take a closer look if I find the time. 
The idea with using the internal version as a fallback sounds good at first glance. Thanks for the suggested code. 
I have only one concern: This could also lead to a situation where noScribe is using a very outdated version of ffmpeg sitting on the user's machine somewhere, instead of the much more recent version packaged with the app. I'll have to think about that.

**Comment by gernophil** [comment]
Can't we just autodownload the most recent version on the first start and regularly check for updates? I would simply store it in the config folder (like the yaml file). I wouldn't use the system's ffmpeg. That could always cause unexpected incompatibilities.

**Comment by tidely** [comment]
The system ffmpeg would get updated routinely by the users package manager. I think we have a few options:

- Always use the system ffmpeg, with no fallback
- Download the fallback if system ffmpeg is unavailable or incompatible
- Always use bundled ffmpeg, update noScribe on every ffmpeg update
- Always used bundled ffmpeg, but update it seperately from noScribe on the users device

Always bundling but then using the system ffmpeg doesn't make sense in my opinion. In my opinion the most optimal solution for the user would be to use the system binary, but incase it's missing or incompatible we download a fallback on first install and update it regularly so we dont need to re-release noScribe on every ffmpeg update. This would mean the user only needs to have an additional ffmpeg installation if they dont have one and it is kept up to date without having to re-install noScribe

**Comment by Lod3** [comment]
This application's advantage over others is the fact that ffmpeg is bundled. 

>Download the fallback if system ffmpeg is unavailable or incompatible

The idea to download ffmpeg's new binaries is a good alternative I think, a piece of software called [Tartube ](https://github.com/axcore/tartube) already does this and it works rather well. 
The issue with that is that the download for ffmpeg is rather large and can takes a long time this is an extra hurdle after the user already downloaded a large executable. These automated downloads and installs can fail and the user is left with questions and a non functioning tool. 

> Always used bundled ffmpeg, but update it seperately from noScribe on the users device
👍 This seems easy for non-technical users. Maybe give an option to ignore the version or even cancel or revert the install if it failed. Because if it breaks a user will reinstall everything or need to. 

> Always use the system ffmpeg, with no fallback

I think relying on the system ffmpeg is a bad idea especially for non-technical users. 
In most cases getting NoScribe to work, you need one installer. Even for the CUDA enabled version, it has a little more requirements but nothing that takes to user to a CLI. This is a strength. 
Most users might not have the technical expertise to install ffmpeg and add it to their path on whatever system, in my experience even if ffmpeg or any other cli tool was installed and added to their path something simple as forgetting to restart their terminal might be a next obstacle. 
Think local history societies, museum/archive employees, researchers, these are the users that benefit ALOT of a tool like this. 
Most of these users I mentioned above are working in a managed IT environment (On Windows, we might all hate Windows, but its the reality) here users need to ask to install a specific piece of software, which in most cases takes over a week or longer. 
IF (because not all IT-departments trust open source software), IT-Department installs NoScribe, this would mean they need to install ffmpeg, IF they followed the requests (which come from a non technical user), "Hey IT I want NoScribe and FFMPEG". What happens is NoScribe is installed Ffmpeg as well but not configured properly, as in not globally adressable/added to path, so the features that rely on ffmpeg dont work. 
Most of the time those users do not create an issue on Github, or even make an account on a website. 

>Always use bundled ffmpeg, update noScribe on every ffmpeg update

From verion 6.0. there where 6 or 7 security issues is that a lot? Is updating NoScribe with those ffmpeg fixes included a lot of work? There is already a feature build in to let the user know there is a new NoScribe verion available. 
Even for non technical users an update notification is still more viable than getting ffmpeg installed AND updating it seperately.

**Comment by tidely** [comment]
I agree with a lot of @Lod3's points. While I think using the system ffmpeg when available and otherwise downloading the binary is the most efficient solution, it might be a bit ambitious. The issue with updating noScribe on each ffmpeg update is the size of the noScribe binary. Since all the whisper models are included in the download, update sizes in the multiple gigabyte range. This discourages users from updating, even though the update may include security fixes. The reality is, that end users often don't care about security that much, and it's up to the developers to make it somewhat mandatory. To prevent ffmpeg from breaking, we could for example have a file in this repository of the maximum ffmpeg version users should use, and we could bump this number up whenever we know that the new binaries are compatible. 

> Can any of the vulnerabilities be exploited by the very fixed command that noScribe is performing using ffmpeg

ffmpeg often doesn't disclose the exact affected component, but for us the most relevant one is probably [this one](https://nvd.nist.gov/vuln/detail/CVE-2024-7055). It affects the decode function of the `libavcodec` component, which is responsible for handling the corresponding audio codec. While the ffmpeg command used in noScribe seems specific, we allow users to normalize *any* audio file to a `.wav` which means almost all audio components of ffmpeg could affect users.

**Comment by kaixxx** [comment]
Thanks for all the great ideas. But I would say, let's not overcomplicate things. Ffmpeg is a reliable piece of software that has been around for ages. And with noScribe, it is not exposed to the outside world. The user decides which audio file to process, no external attacker has direct access. So I still believe it's totally fine to update ffmpeg with each noScribe release and not in between. 

**The real problem is that I have been lazy in the past** and never really checked for new ffmpeg versions when compiling noScribe. I promise to do better in the future! ;) 

It would be great, however, if you could notify us if a serious security vulnerability in ffmpeg arises, so that we can update noScribe. There is a new mechanism in noScribe since v. 0.5 where people are notified about new updates (never used yet). We can also put security warnings in there.
